### ISSUE TITLE
JSX support

### ISSUE THREAD

**Issue by RichardLitt** [feross/standard#47]
So, I'm trying to use standard on .jsx files. And I run into an issue where I can't use `//jscs:disable` and `/* eslint-disable */`, because they still depend on the AST parser and will throw an error if they can't parse the JSX templating. 

Example file: 
```js
, render: function() {
	//jscs:disable
	/*eslint-disable */
    return <div className="form-horizontal">
      {this.renderTextInput('subject', 'Subject')}
      {this.props.email && this.renderTextInput('email', 'Email')}
      {this.renderTextInput('message', 'message')}
    </div>
  //jscs:enable
  /*eslint-enable */
  }
```

What do you think about this? I could potentially use `fs.readFileSync(__dirname, '/form-template.jsx')` and pipe the rendering jsx template through that, but that seems messy. Do you think there would be a way of using standard on these files, otherwise?

**Comment by RichardLitt** [comment]
There ought to be a way to add esprima linting to jscs: cf https://github.com/jscs-dev/node-jscs/issues/708. 

I tried adding `ecmaScript: {..., "jsx": true}` to the `.eslintrc` - didn't seem to work out of the box. More details on eslint and JSX [here](http://eslint.org/docs/configuring/).

**Comment by gaearon** [comment]
Disabling linting for JSX blocks is a mistaken approach. You won't catch potential issues in them, and constants you use there will appear unused to the linter. The right approach is to use JSX-aware linter, or pipe compiled JSX to linter.

**Comment by RichardLitt** [comment]
@gaearon agreed. Do you have any experience making JSX linters work here, like I looked into above? Help would be great.

**Comment by gaearon** [comment]
Wrote an article on teaching ESLint to parse ES6/JSX/some ES7 with Babel a few days ago: https://medium.com/@dan_abramov/lint-like-it-s-2015-6987d44c5b48

It won't help you with `standard` though..

**Comment by Flet** [comment]
Validated using `"jsx":true` in the `"ecmaFeatures"` section of .eslintrc works great, however its not as clear with jscs. I couldn't find any explicit options. 

In searching around their issues, it looks like some folks are suggessing using jsxcs as an alternative. running jsxcs with the current rc file does not show a promising result:
```
Error: Unsupported rules: disallowOperatorBeforeLineBreak, requireSpacesInForStatement
```

So... once #45 is done and only eslint remains, jsx support should be possible.

**Comment by HenrikJoreteg** [comment]
+1 if this gets resolved I can tell folks to use this in my upcoming workshop at FluentConf.

**Comment by feross** [comment]
I'll take another look at adding support for JSX.

Can someone point me to a couple repos with JSX in them? If they already use standard style, even better!

**Comment by RichardLitt** [comment]
@feross Here's mine: https://github.com/BeagleLab/beagle/tree/master/js

Already standardized.

**Comment by HenrikJoreteg** [comment]
@feross also adding you to the little demo app I started on for fluent.

**Comment by Flet** [comment]
Looks like jscs looks jscs should support jsx now, but its not yet been pushed to npm.
https://github.com/jscs-dev/node-jscs/pull/1132

**Comment by Flet** [comment]
OK, so this works, however it depends on jscs in github. The trick is using `esprima-fb`

https://github.com/Flet/standard/commit/5fb221e8b5d8e0a7dae1010bce975bac142b597d

Detailed by jscs contributor here:
https://github.com/jscs-dev/node-jscs/issues/1131

**Comment by Flet** [comment]
old n busted:
>Error: Use JavaScript Standard Style (https://github.com/feross/standard)
  ../beagle/js/views/graphModal.jsx:2:4: Modal is defined but never used
  ../beagle/js/views/graphModal.jsx:3:4: PublicationList is defined but never used
  ../beagle/js/views/graphModal.jsx:7:18: Missing space before function parentheses.
  __../beagle/js/views/graphModal.jsx:9:7: Unexpected token <__
  ../beagle/js/views/graphModal.jsx:26:0: Multiple blank lines not allowed.


new hotness:
>node bin/cmd.js ../beagle/js/views/graphModal.jsx 
Error: Use JavaScript Standard Style (https://github.com/feross/standard)
  ../beagle/js/views/graphModal.jsx:2:4: Modal is defined but never used
  ../beagle/js/views/graphModal.jsx:3:4: PublicationList is defined but never used
  ../beagle/js/views/graphModal.jsx:7:18: Missing space before function parentheses.
  ../beagle/js/views/graphModal.jsx:26:0: Multiple blank lines not allowed.

**Comment by feross** [comment]
Just published `3.1.0` with JSX and React support.

In order to get things like proper detection of used variables inside of JSX, I added [eslint-plugin-react](https://github.com/yannickcr/eslint-plugin-react). Like with standard@1.0.0, let's start with [all React rules](https://github.com/yannickcr/eslint-plugin-react#list-of-supported-rules) enabled and relax them as necessary. Open a new issue if you think a rule is too aggressive.

@HenrikJoreteg, note: I noticed an eslint bug that crashes standard while trying to check your demo app. Here's the eslint issue: https://github.com/eslint/eslint/issues/2060 This isn't related to JSX support :)

I think this issue can be closed now! :+1:

**Comment by feross** [comment]
@Flet thanks for looking into this - looks like we were working in parallel :)

**Comment by HenrikJoreteg** [comment]
woot! thanks @feross @Flet you're awesome.

**Comment by Flet** [comment]
Excellent! Good call on the eslint-plugin-react include :)

**Comment by feross** [comment]
Just released 3.1.1 to automatically lint files that match `**/*.jsx` in addition to the default of `**/*.js`. https://github.com/feross/standard/issues/68

**Comment by feross** [comment]
@HenrikJoreteg The eslint issue I mentioned (https://github.com/eslint/eslint/issues/2060) is fixed now :)

**Comment by HenrikJoreteg** [comment]
just noticed that, thanks @feross :)

**Comment by RichardLitt** [comment]
Works for me! Thanks all!

**Comment by anthonybrown** [comment]
Works great for my React projects!
